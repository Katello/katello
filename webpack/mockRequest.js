import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import nock from 'nock';

// TODO: figure out way to reuse this from foreman
export const mock = new MockAdapter(axios);
const methods = {
  GET: 'onGet',
  POST: 'onPost',
  PUT: 'onPut',
  DELETE: 'onDelete',
};

const errorResponse = msg => ({ displayMessage: msg });

export const mockRequest = ({
  method = 'GET',
  url,
  data,
  status = 200,
  response = null,
}) => mock[methods[method.toUpperCase()]](url, data).reply(status, response);

export const mockErrorRequest = ({
  status = 500,
  ...options
}) => mockRequest({
  response: errorResponse(`Request failed with status code ${status}`),
  status,
  ...options,
});

export const mockReset = () => mock.reset();

// Using the library 'nock' as it matches actual network requests rather than mock another
// library. This is helpful when the request is not coming from Katello. For example, axios
// called within Katello can be mocked with axios-mock-adapter or similar, but a http request
// made by axios that is coming from Foreman cannot be mocked by axios-mock-adapter or a
// jest mock within Katello. So to do this, we can mock the request a level deeper within
// nodejs by using nock.
export const nockInstance = nock('http://localhost');

// Calling .done() with nock asserts that the request was fufilled. We use a timeout to ensure
// that the component has set up and made the request before the assertion is made.
export const assertNockRequest = (scope, timeout = 2000) => {
  setTimeout(() => {
    scope.done();
  }, timeout);
};
