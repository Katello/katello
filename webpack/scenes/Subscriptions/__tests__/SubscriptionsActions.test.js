import thunk from 'redux-thunk';
import Immutable from 'seamless-immutable';
import configureMockStore from 'redux-mock-store';
import { testActionSnapshotWithFixtures } from 'react-redux-test-utils';
import { mockRequest, mockErrorRequest, mockReset } from '../../../mockRequest';
import {
  requestSuccessResponse,
  requestSuccessResponseWithRHSubscriptions,
  successActions,
  successActionsWithQuantityLoad,
  failureActions,
  updateQuantitySuccessActions,
  updateQuantityFailureActions,
  poolsUpdate,
  loadQuantitiesFailureActions,
  loadQuantitiesSuccessActions,
  quantitiesRequestSuccessResponse,
  loadTableColumnsSuccessAction,
} from './subscriptions.fixtures';
import {
  loadSubscriptions,
  updateQuantity,
  loadAvailableQuantities,
  loadTableColumns,
  updateSearchQuery,
  openDeleteModal,
  closeDeleteModal,
  openTaskModal,
  closeTaskModal,
  disableDeleteButton,
  enableDeleteButton,
} from '../SubscriptionActions';

const mockStore = configureMockStore([thunk]);
const store = mockStore({ subscriptions: Immutable({}) });

afterEach(() => {
  store.clearActions();
  mockReset();
});

describe('subscription actions', () => {
  describe('loadSubscriptions', () => {
    it(
      'creates SUBSCRIPTIONS_REQUEST and then fails with 422',
      async () => {
        mockErrorRequest({
          url: '/katello/api/v2/subscriptions',
          status: 422,
        });
        await store.dispatch(loadSubscriptions());
        expect(store.getActions()).toEqual(failureActions);
      },
    );
    it(
      'creates SUBSCRIPTIONS_REQUEST and ends with success',
      async () => {
        mockRequest({
          url: '/katello/api/v2/subscriptions',
          response: requestSuccessResponse,
        });
        await store.dispatch(loadSubscriptions());
        expect(store.getActions()).toEqual(successActions);
      },
    );
    it(
      'creates SUBSCRIPTIONS_REQUEST and triggers loadAvailableQuantities when there is some RH subscription',
      async () => {
        mockRequest({
          url: '/katello/api/v2/subscriptions',
          response: requestSuccessResponseWithRHSubscriptions,
        });
        await store.dispatch(loadSubscriptions());
        expect(store.getActions()).toEqual(successActionsWithQuantityLoad);
      },
    );
  });

  describe('updateQuantity', () => {
    it(
      'creates UPDATE_QUANTITY_REQUEST and then fails with 422',
      async () => {
        mockErrorRequest({
          method: 'PUT',
          url: '/katello/api/v2/organizations/1/upstream_subscriptions',
          data: { pools: poolsUpdate },
          status: 422,
        });
        await store.dispatch(updateQuantity(poolsUpdate));
        expect(store.getActions()).toEqual(updateQuantityFailureActions);
      },
    );
    it(
      'creates UPDATE_QUANTITY_REQUEST and ends with success',
      async () => {
        mockRequest({
          method: 'PUT',
          url: '/katello/api/v2/organizations/1/upstream_subscriptions',
          data: { pools: poolsUpdate },
          response: requestSuccessResponse,
        });
        await store.dispatch(updateQuantity(poolsUpdate));
        expect(store.getActions()).toEqual(updateQuantitySuccessActions);
      },
    );
  });

  describe('loadAvailableQuantities', () => {
    const data = { pool_ids: [5] };

    it(
      'creates SUBSCRIPTIONS_QUANTITIES_REQUEST and then fails with 500',
      async () => {
        mockErrorRequest({
          method: 'GET',
          url: '/katello/api/v2/organizations/1/upstream_subscriptions',
          data,
          status: 500,
        });
        await store.dispatch(loadAvailableQuantities());
        expect(store.getActions()).toEqual(loadQuantitiesFailureActions);
      },
    );
    it(
      'creates SUBSCRIPTIONS_QUANTITIES_REQUEST and ends with success',
      async () => {
        mockRequest({
          method: 'GET',
          url: '/katello/api/v2/organizations/1/upstream_subscriptions',
          data,
          response: quantitiesRequestSuccessResponse,
        });
        await store.dispatch(loadAvailableQuantities());
        expect(store.getActions()).toEqual(loadQuantitiesSuccessActions);
      },
    );
  });
  describe('loadTableColumns', () => {
    it(
      'loads table columns',
      () => {
        store.dispatch(loadTableColumns());
        expect(store.getActions()).toEqual(loadTableColumnsSuccessAction);
      },
    );
  });

  describe('deleteModal', () => testActionSnapshotWithFixtures({
    'it should open delete modal': () => openDeleteModal(),
    'it should close delete modal': () => closeDeleteModal(),
  }));

  describe('searchQuery', () => testActionSnapshotWithFixtures({
    'it should update the search-query': () => updateSearchQuery('some-query'),
  }));

  describe('taskModal', () => testActionSnapshotWithFixtures({
    'it should open task modal': () => openTaskModal(),
    'it should close task modal': () => closeTaskModal(),
  }));

  describe('deleteButtonDisabled', () => testActionSnapshotWithFixtures({
    'it should disable the delete button': () => disableDeleteButton(),
    'it should enable the delete button': () => enableDeleteButton(),
  }));
});
