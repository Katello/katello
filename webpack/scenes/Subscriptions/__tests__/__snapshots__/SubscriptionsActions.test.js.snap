// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`subscription actions deleteButtonDisabled it should disable the delete button 1`] = `
Object {
  "type": "SUBSCRIPTIONS_DISABLE_DELETE_BUTTON",
}
`;

exports[`subscription actions deleteButtonDisabled it should enable the delete button 1`] = `
Object {
  "type": "SUBSCRIPTIONS_ENABLE_DELETE_BUTTON",
}
`;

exports[`subscription actions deleteModal it should close delete modal 1`] = `
Object {
  "type": "SUBSCRIPTIONS_CLOSE_DELETE_MODAL",
}
`;

exports[`subscription actions deleteModal it should open delete modal 1`] = `
Object {
  "type": "SUBSCRIPTIONS_OPEN_DELETE_MODAL",
}
`;

exports[`subscription actions searchQuery it should update the search-query 1`] = `
Object {
  "payload": "some-query",
  "type": "SUBSCRIPTIONS_UPDATE_SEARCH_QUERY",
}
`;

exports[`subscription actions taskModal it should close task modal 1`] = `
Object {
  "type": "SUBSCRIPTIONS_CLOSE_TASK_MODAL",
}
`;

exports[`subscription actions taskModal it should open task modal 1`] = `
Object {
  "type": "SUBSCRIPTIONS_OPEN_TASK_MODAL",
}
`;
